httpcore/_backends/curio.py

```diff
@@ -0,0 +1,214 @@
+from .base import AsyncBackend, AsyncLock, AsyncSemaphore, AsyncSocketStream
+
+logger = get_logger("curio_backend")
+
+one_day_in_seconds = 60 * 60 * 24
+
+
+async def wrap_ssl_client(
+    sock: curio.io.Socket,
+    ssl_context: SSLContext,
+    server_hostname: bytes,
+) -> curio.io.Socket:
+    kwargs = {
+        "server_hostname": server_hostname,
+        "do_handshake_on_connect": sock._socket.gettimeout() != 0.0,
+    }
+
+    socket = curio.io.Socket(ssl_context.wrap_socket(sock._socket, **kwargs))
+    await socket.do_handshake()
+
+    return socket
+
+
+def convert_timeout(value: Optional[float]) -> int:
+    return int(value) if value is not None else one_day_in_seconds
+
+
+class Lock(AsyncLock):
+    def __init__(self) -> None:
+        self._lock = curio.Lock()
+
+    async def acquire(self) -> None:
+        await self._lock.acquire()
+
+    async def release(self) -> None:
+        await self._lock.release()
+
+
+class Semaphore(AsyncSemaphore):
+    def __init__(self, max_value: int, exc_class: type) -> None:
+        self.max_value = max_value
+        self.exc_class = exc_class
+
+    @property
+    def semaphore(self) -> curio.Semaphore:
+        if not hasattr(self, "_semaphore"):
+            self._semaphore = curio.Semaphore(value=self.max_value)
+        return self._semaphore
+
+    async def acquire(self, timeout: float = None) -> None:
+        exc_map: Dict[Type[Exception], Type[Exception]] = {
+            curio.TaskTimeout: TimeoutException,
+        }
+        acquire_timeout: int = convert_timeout(timeout)
+
+        with map_exceptions(exc_map):
+            return await curio.timeout_after(acquire_timeout, self.semaphore.acquire())
+
+    async def release(self) -> None:
+        await self.semaphore.release()
+
+
+class SocketStream(AsyncSocketStream):
+    def __init__(self, socket: curio.io.Socket) -> None:
+        self.read_lock = curio.Lock()
+        self.write_lock = curio.Lock()
+        self.socket = socket
+        self.stream = socket.as_stream()
+
+    def get_http_version(self) -> str:
+        ident: Optional[str] = "http/1.1"
+
+        if hasattr(self.socket, "_socket"):
+            raw_socket: Union[SSLSocket, socket.socket] = self.socket._socket
+
+            if isinstance(raw_socket, SSLSocket):
+                ident = raw_socket.selected_alpn_protocol()
+
+        return "HTTP/2" if ident == "h2" else "HTTP/1.1"
+
+    async def start_tls(
+        self, hostname: bytes, ssl_context: SSLContext, timeout: TimeoutDict
+    ) -> "AsyncSocketStream":
+        connect_timeout = convert_timeout(timeout.get("connect"))
+        exc_map = {
+            curio.TaskTimeout: ConnectTimeout,
+            curio.CurioError: ConnectError,
+            OSError: ConnectError,
+        }
+
+        with map_exceptions(exc_map):
+            wrapped_sock = await curio.timeout_after(
+                connect_timeout,
+                wrap_ssl_client(self.socket, ssl_context, hostname),
+            )
```

Comment:
Since the `wrap_ssl_client` function is only used here, should we consider just moving code in here…? (Benefit: it's clearer that the TLS handshake is the only bit that might cause a timeout.)

```suggestion
            wrapped_socket = curio.io.Socket(
                ssl_context.wrap_socket(
                    self.socket._socket,   
                    server_hostname=hostname,
                    do_handshake_on_connect=self.socket._socket.gettimeout() != 0,
                )
            )

            await curio.timeout_after(
                connect_timeout,
                wrapped_sock.do_handshake(),
            )
```

Reply:
Removed `wrap_ssl_client`, and done the suggested changes.

By the way I set `do_handshake_on_connect = False` as we do the handshake right after the socket creation